fe2-android/app/src/androidTest/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivityTest.java

```diff
@@ -0,0 +1,182 @@
+import dagger.hilt.android.testing.HiltAndroidTest;
+import io.reactivex.subjects.BehaviorSubject;
+
+/** Test of the LaoDetail Activity */
+@HiltAndroidTest
+public class LaoDetailActivityTest {
+
+  private static final Lao LAO = new Lao("LAO", Base64DataUtils.generatePublicKey(), 10223421);
+  private static final String LAO_ID = LAO.getId();
+
+  @Inject LAORequestFactory requestFactory;
+  @Inject Gson gson;
+
+  @BindValue @Mock LAORepository laoRepository;
+
+  // Hilt rule
+  private final HiltAndroidRule hiltAndroidRule = new HiltAndroidRule(this);
+  // Setup rule, used to setup things before the activity is started
+  private final TestRule setupRule =
+      new ExternalResource() {
+        @Override
+        protected void before() {
+          hiltAndroidRule.inject();
+
+          when(laoRepository.getLaoObservable(anyString()))
+              .thenReturn(BehaviorSubject.createDefault(LAO));
+        }
+      };
+  // Activity scenario rule that starts the activity.
+  // It creates a LaoDetailActivity with set extras such that the LAO is used
+  public ActivityScenarioRule<LaoDetailActivity> activityScenarioRule =
+      new ActivityScenarioRule<>(
+          IntentUtils.createIntent(
+              LaoDetailActivity.class,
+              new BundleBuilder()
+                  .putString(laoIdExtra(), LAO_ID)
+                  .putString(fragmentToOpenExtra(), laoDetailValue())
+                  .build()));
+
+  @Rule
+  public final RuleChain rule =
+      RuleChain.outerRule(MockitoJUnit.testRule(this))
+          .around(hiltAndroidRule)
+          .around(setupRule)
+          .around(activityScenarioRule);
+
+  @Test
+  public void homeButtonOpensHome() {
+    homeButton().perform(click());
+
+    assertEquals(
+        HomeActivity.class.getName(),
+        activityScenarioRule
+            .getScenario()
+            .getResult()
+            .getResultData()
+            .getComponent()
+            .getClassName());
+  }
+
+  @Test
+  public void identityButtonOpensIdentityTab() {
+    identityButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(identityFragmentId()))));
+  }
+
+  @Test
+  public void showPropertyButtonShowsConnectQRCode() {
+    showPropertiesButton().perform(click());
+
+    propertiesLayout().check(matches(isDisplayed()));
+
+    String expectedQRCode = gson.toJson(new ConnectToLao(requestFactory.getUrl(), LAO_ID));
+    connectQrCode().check(matches(withQrCode(expectedQRCode)));
+  }
+
+  @Test
+  public void witnessButtonShowsWitnessTab() {
+    witnessButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(witnessFragmentId()))));
+  }
+
+  // Matches an ImageView containing a QRCode with expected content
+  private Matcher<? super View> withQrCode(String expectedContent) {
+    return new TypeSafeMatcher<View>() {
+      @Override
+      protected boolean matchesSafely(View item) {
+        if (item instanceof ImageView) {
```

Comment:
could we do the following ? 
```suggestion
    return new TypeSafeMatcher<ImageView>() {
      @Override
      protected boolean matchesSafely(ImageView item) {
```

Reply:
This is not possible because TypeSafeMatcher<ImageView> is not a Matcher<? super View>
But I found another way :  BoundedMatcher.